// @project-serum/serum/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var market_1 = require("./market");
Object.defineProperty(exports, "Market", { enumerable: true, get: function () { return market_1.Market; } });
Object.defineProperty(exports, "Orderbook", { enumerable: true, get: function () { return market_1.Orderbook; } });
Object.defineProperty(exports, "OpenOrders", { enumerable: true, get: function () { return market_1.OpenOrders; } });
var instructions_1 = require("./instructions");
Object.defineProperty(exports, "DexInstructions", { enumerable: true, get: function () { return instructions_1.DexInstructions; } });
Object.defineProperty(exports, "decodeInstruction", { enumerable: true, get: function () { return instructions_1.decodeInstruction; } });
Object.defineProperty(exports, "SETTLE_FUNDS_BASE_WALLET_INDEX", { enumerable: true, get: function () { return instructions_1.SETTLE_FUNDS_BASE_WALLET_INDEX; } });
Object.defineProperty(exports, "SETTLE_FUNDS_QUOTE_WALLET_INDEX", { enumerable: true, get: function () { return instructions_1.SETTLE_FUNDS_QUOTE_WALLET_INDEX; } });
Object.defineProperty(exports, "NEW_ORDER_OPEN_ORDERS_INDEX", { enumerable: true, get: function () { return instructions_1.NEW_ORDER_OPEN_ORDERS_INDEX; } });
Object.defineProperty(exports, "NEW_ORDER_OWNER_INDEX", { enumerable: true, get: function () { return instructions_1.NEW_ORDER_OWNER_INDEX; } });
var fees_1 = require("./fees");
Object.defineProperty(exports, "getFeeTier", { enumerable: true, get: function () { return fees_1.getFeeTier; } });
Object.defineProperty(exports, "getFeeRates", { enumerable: true, get: function () { return fees_1.getFeeRates; } });
Object.defineProperty(exports, "supportsSrmFeeDiscounts", { enumerable: true, get: function () { return fees_1.supportsSrmFeeDiscounts; } });
var tokens_and_markets_1 = require("./tokens_and_markets");
Object.defineProperty(exports, "TOKEN_MINTS", { enumerable: true, get: function () { return tokens_and_markets_1.TOKEN_MINTS; } });
Object.defineProperty(exports, "MARKETS", { enumerable: true, get: function () { return tokens_and_markets_1.MARKETS; } });
Object.defineProperty(exports, "getLayoutVersion", { enumerable: true, get: function () { return tokens_and_markets_1.getLayoutVersion; } });
var queue_1 = require("./queue");
Object.defineProperty(exports, "decodeEventQueue", { enumerable: true, get: function () { return queue_1.decodeEventQueue; } });
Object.defineProperty(exports, "decodeRequestQueue", { enumerable: true, get: function () { return queue_1.decodeRequestQueue; } });
Object.defineProperty(exports, "REQUEST_QUEUE_LAYOUT", { enumerable: true, get: function () { return queue_1.REQUEST_QUEUE_LAYOUT; } });
Object.defineProperty(exports, "EVENT_QUEUE_LAYOUT", { enumerable: true, get: function () { return queue_1.EVENT_QUEUE_LAYOUT; } });
exports.TokenInstructions = __importStar(require("./token-instructions"));
